### ISSUE TITLE
On the interaction between DB and repositories

### ISSUE THREAD

**Issue by alalazo** [spack/spack#3515]
**TLDR** Below I am trying to make the point on a few bugs that led to many open discussions, but that in my opinion share the same common problems. I hope this will help in the analysis and that will start a discussion on how to solve them.

--------

There are currently a lot of discussions on issues arising from:

1. installing software from a repo
2. changing or deleting the `package.py` for the software above
3. manipulating the installation done at point 1.

A list of such discussions is #2911, #3350, #3476, #3501, #3502 and I am pretty sure I still forget some.

Examples of unwanted behaviors may vary from having errors like:
```console
$ spack module refresh --module-type lmod

...
  File "/blues/gpfs/home/software/spack-0.10.0/lib/spack/spack/directory_layout.py", line 42, in _check_concrete
    raise ValueError('Specs passed to a DirectoryLayout must be concrete!')
ValueError: Specs passed to a DirectoryLayout must be concrete!
```
when trying to refresh module files to being unable to uninstall packages if in the meanwhile you deleted the corresponding `package.py` file.

### What I think is the root cause of all the issues

From what I can tell the root of these problems is **an inconsistent state of the spec object retrieved from the DB**. I'll try to clarify that.

When doing something like:

```python
specs = spack.store.db.query()
```

the operations we are performing are:

- reading the DB of installed specs
- marking them concrete

Note that installed specs are by definition concrete *if compared to the `package.py` files that were used when they were installed*. Now, suppose I have the following line of code:
```python
pkg = specs[0].package
```
What I would expect for `pkg` is to be conforming with the state of things at the time I installed `specs[0]`. Inspecting the `Spec.package` property though:
```python
@property
def package(self):
    return spack.repo.get(self)
```
one can see that instead you get back **the current state of the package**. If in the meanwhile a variant was added, that variant will appear in `pkg`. If the package was deleted from the repo, the corresponding file will not be found. **This is in my opinion the core issue of all the above discussions**.

### Things I find strange in the spec/package interactions

On top of the issue above, I also noticed a strange behavior in spec/package interaction that I first reported [here](https://github.com/LLNL/spack/issues/2911#issuecomment-274635426):
```python
# a is m4@1.4.18%gcc@4.8+sigsegv arch=linux-ubuntu14-x86_64 ^libsigsegv@2.10%gcc@4.8 arch=linux-ubuntu14-x86_64
a = specs[1]
a.concrete, a['libsigsegv'].concrete
Out[3]: (True, True)
a = specs[1].package.spec
a.concrete, a['libsigsegv'].concrete
Out[5]: (True, False)
```
I still think it's a bug, and probably it has to do with `Spec.copy` and its helper methods. Intuitively I would expect that:
```python
spec == spec.package.spec
spec.package.spec == spec.package.spec.package.spec
```
but apparently only the second line is always true, the first one is not. From what I can tell we need tests and fixes for this too, but I would like somebody else (@tgamblin, @becker33, @scheibelp) to confirm this.

**Comment by alalazo** [comment]
For the record, I think we should really:

1. Get rid of all the direct usages of global variables like `spack.repo` in classes or functions. There should be an indirection layer to make it easy to plug-in something different.
2. Make repository with the same namespace extend each other, with a semantic that resembles dictionary updates
3. Change the mechanism used to detect if a spec is virtual or not

[Here](https://github.com/LLNL/spack/pull/3350#issuecomment-284490656) there are a few more details why I think the points above are necessary.

**Comment by scheibelp** [comment]
>  but I would like somebody to confirm this.

I can definitely confirm that spec and spec.package.spec are not always consistent. In fact there is special logic in ```build_environment.setup_package``` to correct this:

```
for s in pkg.spec.traverse():
  s.package.spec = s
```

I have previously suspected ```repository.Repo.get```; FWIW my initial guess would be that ```hash(spec)``` is getting thrown off because deptypes are not copied properly (something in the logic of ```Spec._dup``` and ```Spec._dup_deps```). I have encountered at least one case where the internals of copied specs do not precisely match, although in this case (https://github.com/LLNL/spack/pull/2928) it was more a matter of cached variables not being copied, which should be fine.

**Comment by alalazo** [comment]
Less of an issue lately, but ♻ deps between Spec and PackageBase are not entirely removed.
